scylla/src/transport/query_result.rs

```diff
@@ -0,0 +1,513 @@
+        self.specs.iter().map(ColumnSpecView::new_from_column_spec)
+    }
+}
+
+/// Result of a single request to the database. It represents any kind of Result frame.
+///
+/// The received rows and metadata, which are present if the frame is of Result:Rows kind,
+/// are kept in a raw binary form. To deserialize and access them, this struct works
+/// in tandem with [`RowsDeserializer`] struct, which borrows from [`QueryResult`].
+/// By borrowing, [`RowsDeserializer`] can avoid heap allocations of metadata strings,
+/// borrowing them from the Result frame instead.
+/// To create a [`RowsDeserializer`], use [`QueryResult::rows_deserializer`] method.
+/// Upon creation, [`RowsDeserializer`] deserializes result metadata and allocates it,
+/// so this should be considered a moderately costly operation and performed only once.
+///
+/// NOTE: this is a result of a single CQL request. If you use paging for your query,
+/// this will contain exactly one page.
+#[derive(Debug)]
+pub struct QueryResult {
+    raw_rows: Option<RawRows>,
+    tracing_id: Option<Uuid>,
+    warnings: Vec<String>,
+}
+
+impl QueryResult {
+    pub(crate) fn new(
+        raw_rows: Option<RawRows>,
+        tracing_id: Option<Uuid>,
+        warnings: Vec<String>,
+    ) -> Self {
+        Self {
+            raw_rows,
+            tracing_id,
+            warnings,
+        }
+    }
+
+    // Preferred to implementing Default, because users shouldn't be able to create
+    // an empty QueryResult.
+    //
+    // For now unused, but it will be used once Session's API is migrated
+    // to the new QueryResult.
+    #[allow(dead_code)]
+    pub(crate) fn mock_empty() -> Self {
+        Self {
+            raw_rows: None,
+            tracing_id: None,
+            warnings: Vec::new(),
+        }
+    }
+
+    /// Warnings emitted by the database.
+    #[inline]
+    pub fn warnings(&self) -> impl Iterator<Item = &str> {
+        self.warnings.iter().map(String::as_str)
+    }
+
+    /// Tracing ID associated with this CQL request.
+    #[inline]
+    pub fn tracing_id(&self) -> Option<Uuid> {
+        self.tracing_id
+    }
+
+    /// Returns a bool indicating the current response is of Rows type.
+    #[inline]
+    pub fn is_rows(&self) -> bool {
+        self.raw_rows.is_some()
+    }
+
+    /// Returns `Ok` for a request's result that shouldn't contain any rows.\
+    /// Will return `Ok` for `INSERT` result, but a `SELECT` result, even an empty one, will cause an error.\
+    /// Opposite of [`rows_deserializer()`](QueryResult::rows_deserializer).
+    #[inline]
+    pub fn result_not_rows(&self) -> Result<(), ResultNotRowsError> {
+        match &self.raw_rows {
+            Some(_) => Err(ResultNotRowsError),
+            None => Ok(()),
+        }
+    }
+
+    /// Creates a lifetime-bound [`RowsDeserializer`] to enable deserializing rows contained
+    /// in this [`QueryResult`]'s frame. Deserializes result metadata and allocates it,
```

Comment:
Even for the borrowed version, `ColumnType`, being a recursive type, must be heap-allocated (using Boxes) in case of some of its variants. So for each column we potentially have a number of allocations, too.